packages/calypso-products/src/is-difm-product.ts

```diff
@@ -1,8 +1,7 @@
+import { camelOrSnakeSlug } from './camel-or-snake-slug';
 import { WPCOM_DIFM_LITE } from './constants';
-import { formatProduct } from './format-product';
+import type { WithCamelCaseSlug, WithSnakeCaseSlug } from './types';
 
-export function isDIFMProduct( product: { product_slug?: string; productSlug?: string } ): boolean {
-	product = formatProduct( product );
-
-	return product.product_slug === WPCOM_DIFM_LITE;
+export function isDIFMProduct( product: WithCamelCaseSlug | WithSnakeCaseSlug ): boolean {
```

Comment:
This is probably not worth acting on, at least in this PR, but it feels like an implementation detail leaking out that this file needs to know about camel case and snake case slugs. For instance, `types` could define a type synonym like `WithProductSlug = WithCamelCaseSlug | WithSnakeCaseSlug`, then we could have a utility function like `getProductSlug` that hides the fact that that the key for that slug might be in camelCase or snake_case from consumers.

Reply:
Yes, and I thought of that but I kind of liked the explicit nature of specifying both types... maybe it's not important since TypeScript _should_ provide guarantees either way.

Reply:
I have a plan to try to replace many of these simpler functions with something like `doesProductMatchSlug(product: WithProductSlug, slug: string)` and maybe that would be a good place to make this improvement. The main goal of these PRs so far has just been to convert everything to TS and get rid of the ambiguity of `formatProduct`.

Reply:
I always get tripped up by the fact that type equality in TS is structural rather than nominal, so maybe there are fewer safety benefits to introducing an "indirect" type like `WithProductSlug`. As I see it the fact that we sometimes end up with `productSlug` and other times with `product_slug`, when they are indistinguishable semantically, is a bug, and tech debt that will probably never be paid down since it is so pervasive.

To explain why I think hiding the snake/camel distinction is preferable - because of the ambiguity coming from the API, "product slug" is not actually a syntactic property of the object we can just read, but instead a semantic property that has to be computed (trivially, but still). Then the type `WithCamelCaseSlug | WithSnakeCaseSlug` is encoding _how_ the property is computed, when consumers like `isDIFMProduct` only care that the property _can be_ computed. This is where the nominal checker trips my structural intuition; I'm not sure if this is a meaningful distinction to make in TS without making more intrusive changes.

At any rate my comment is mostly navel gazing and not a blocker. I also think your `doesProductMatchSlug` plan will be a step in the right direction.

Reply:
> as I see it the fact that we sometimes end up with productSlug and other times with product_slug, when they are indistinguishable semantically, is a bug,

I agree. That's actually what I meant about liking the explicit nature of having both types written out every time; it might help convince devs in the future to use just one or the other rather than superstitiously allowing both by copy/pasting what has come before. 

The first few of the changes in this series did, in fact, limit the calls to either snake_case or camelCase only, but these later functions are called in so many non-TypeScript places that it was impossible to safely and easily verify the data they actually pass; I decided that removing formatProduct and adding types by itself had enough value for now.

> I always get tripped up by the fact that type equality in TS is structural rather than nominal...
> "product slug" is not actually a syntactic property of the object we can just read, but instead a semantic property that has to be computed (trivially, but still). Then the type WithCamelCaseSlug | WithSnakeCaseSlug is encoding how the property is computed,

You're totally right. I think TS is actually _both_ structural and nominal, depending on usage. Technically it's structural, sure, but good types encourage a object-oriented way of viewing data and hiding implementation detail. Again, in my first few PRs like this one I used `Product` for many of these arguments because that carries much more semantic meaning. However, as I wasn't able to figure out all the possible data types being passed in, I eventually gave up and went for the structural approach. :-/